### PR TITLE
Gamma fit for discpot

### PR DESCRIPTION
**Pull request by chiarabellenghi**

### CODE REVIEW

File: skyllh/core/analysis_utils.py [hunk 1]
```diff
@@ -83,6 +85,80 @@ def calculate_pval_from_trials(
...
+def truncated_gamma_logpdf(
+        a, scale, eta, ts_above_eta, N_above_eta):
+    """Calculates the -log(likelihood) of a sample of random numbers
+    generated from a gamma pdf truncated from below at x=eta.
+
+    Parameters
+    ----------
+    a : float
+        Shape parameter.
+    scale : float
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:97]:
```suggestion
    scale : float
        Scale parameter.
```

File: skyllh/core/analysis_utils.py [hunk 2]
```diff
@@ -83,6 +85,80 @@ def calculate_pval_from_trials(
...
+    eta : float
+        Test-statistic value at which the gamma function is truncated
+        from below.
+    ts_above_eta : (n_trials,)-shaped 1D ndarray
+        The ndarray holding the test-statistic values falling in
+        the truncated gamma pdf..
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:103]:
```suggestion
        the truncated gamma pdf.
```

File: skyllh/core/analysis_utils.py [hunk 3]
```diff
@@ -83,6 +85,80 @@ def calculate_pval_from_trials(
...
+        gamma pdf.
+        
+    Returns
+    -------
+    -logl : float
+    """
+    c0 = 1. - gamma.cdf(eta, a=a, scale=scale)
+    c0 = 1./c0
+    logl = N_above_eta * np.log(c0) + np.sum(gamma.logpdf(ts_above_eta,
+                                                a=a, scale=scale))
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:115]:
```suggestion
    logl = N_above_eta*np.log(c0) + np.sum(gamma.logpdf(ts_above_eta,
                                                        a=a, scale=scale))
```

File: skyllh/core/analysis_utils.py [hunk 4]
```diff
@@ -83,6 +85,80 @@ def calculate_pval_from_trials(
...
+        Approximate estimate of the median of the test-statistic
+        distribution.
+
+    Returns
+    -------
+    critical_ts : float
+    """
+    Ntot = len(ts)
+    N = len(ts[ts > xi])
+    alpha = N / Ntot
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:143]:
```suggestion
    alpha = N/Ntot
```

File: skyllh/core/analysis_utils.py [hunk 5]
```diff
@@ -83,6 +85,80 @@ def calculate_pval_from_trials(
...
+
+    ts_eta = ts[ts > eta]
+    N_prime = len(ts_eta)
+    alpha_prime = N_prime / N
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:147]:
```suggestion
    alpha_prime = N_prime/N
```

File: skyllh/core/analysis_utils.py [hunk 6]
```diff
@@ -83,6 +85,80 @@ def calculate_pval_from_trials(
...
+
+    obj = lambda x: truncated_gamma_logpdf(x[0], x[1], eta=eta,
+                                            ts_above_eta=ts_eta,
+                                            N_above_eta=N_prime)
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:151]:
```suggestion
                                           ts_above_eta=ts_eta,
                                           N_above_eta=N_prime)
```

File: skyllh/core/analysis_utils.py [hunk 7]
```diff
@@ -83,6 +85,80 @@ def calculate_pval_from_trials(
...
+    x0 = [0.75, 1.8] # starting values
+    bounds = [ [0.1, 10], [0.1, 10] ] # ranges for fitter
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:153]:
```suggestion
    bounds = [[0.1, 10], [0.1, 10]] # ranges for fitter
```

File: skyllh/core/analysis_utils.py [hunk 8]
```diff
@@ -83,6 +85,80 @@ def calculate_pval_from_trials(
...
+    alpha = N / Ntot
+
+    ts_eta = ts[ts > eta]
+    N_prime = len(ts_eta)
+    alpha_prime = N_prime / N
+
+    obj = lambda x: truncated_gamma_logpdf(x[0], x[1], eta=eta,
+                                            ts_above_eta=ts_eta,
+                                            N_above_eta=N_prime)
+    x0 = [0.75, 1.8] # starting values
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:152]:
```suggestion
    x0 = [0.75, 1.8]  # Initial values of function parameters.
```

File: skyllh/core/analysis_utils.py [hunk 9]
```diff
@@ -83,6 +85,80 @@ def calculate_pval_from_trials(
...
+    r = minimize(obj, x0, bounds=bounds)
+    pars = r.x
+    
+    interval = gamma.interval(1-h0_ts_quantile, a=pars[0], scale=pars[1])
+    critical_ts = interval[1]
+
+    return critical_ts
+
 def polynomial_fit(
         ns, p, p_weight, deg, p_thr):
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:163]:
```suggestion
def polynomial_fit(ns, p, p_weight, deg, p_thr):
```

File: skyllh/core/analysis_utils.py [hunk 10]
```diff
@@ -209,12 +309,21 @@ def estimate_mean_nsignal_for_ts_quantile(
...
         'Min / Max h0 TS value: %e / %e',
         np.min(h0_ts_vals), np.max(h0_ts_vals))
-
-    c = np.percentile(h0_ts_vals, (1 - h0_ts_quantile) * 100)
+    
+    # If h0_ts_quantile is larger than 1.e3, compute the critical ts value
+    # directly from trials; otherwise calculate it from the gamma function
+    # fitted to the ts distribution.
+    if(h0_ts_quantile >= 1.e-3):
+        c = np.percentile(h0_ts_vals, (1 - h0_ts_quantile) * 100)
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:317]:
```suggestion
        c = np.percentile(h0_ts_vals, (1 - h0_ts_quantile)*100)
```

File: skyllh/core/analysis_utils.py [hunk 11]
```diff
@@ -209,12 +311,21 @@ def estimate_mean_nsignal_for_ts_quantile(
...
+    else:
+        c = calculate_critical_ts_from_gamma(h0_ts_vals, h0_ts_quantile)
     logger.debug(
         'Critical ts value for bkg ts quantile %g: %e',
         h0_ts_quantile, c)
 
+    if(pathfilename is not None):
+        np.save(pathfilename, h0_ts_vals)
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:327]:
```suggestion
        os.makedirs(os.path.dirname(pathfilename), exist_ok=True)
        np.save(pathfilename, h0_ts_vals)
```

File: skyllh/core/analysis_utils.py [hunk 12]
```diff
@@ -532,6 +643,9 @@ def estimate_sensitivity(
         The possible parent ProgressBar instance.
     tl: instance of TimeLord | None
         The optional TimeLord instance that should be used to collect timing
+    pathfilename : string | None
+        Trial data file path including the filename.
+        If set to None, trials won't be saved.
         information about this function.
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:649]:
```suggestion
        The optional TimeLord instance that should be used to collect timing
        information about this function.
    pathfilename : string | None
        Trial data file path including the filename.
        If set to None, trials won't be saved.
```

File: skyllh/core/analysis_utils.py [hunk 13]
```diff
@@ -83,6 +87,80 @@ def calculate_pval_from_trials(
...
+
+    obj = lambda x: truncated_gamma_logpdf(x[0], x[1], eta=eta,
+                                            ts_above_eta=ts_eta,
+                                            N_above_eta=N_prime)
+    x0 = [0.75, 1.8] # starting values
+    bounds = [ [0.1, 10], [0.1, 10] ] # ranges for fitter
+    r = minimize(obj, x0, bounds=bounds)
+    pars = r.x
+    
+    interval = gamma.interval(1-h0_ts_quantile, a=pars[0], scale=pars[1])
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:159]:
```suggestion
    interval = gamma.interval(1 - h0_ts_quantile, a=pars[0], scale=pars[1])
```

File: skyllh/core/analysis_utils.py [hunk 14]
```diff
@@ -83,8 +87,82 @@ def calculate_pval_from_trials(
...
+    -------
+    critical_ts : float
+    """
+    Ntot = len(ts)
+    N = len(ts[ts > xi])
+    alpha = N/Ntot
+
+    ts_eta = ts[ts > eta]
+    N_prime = len(ts_eta)
+    alpha_prime = N_prime/N
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:150]:
```suggestion
    ts_eta = ts[ts > eta]
    N_prime = len(ts_eta)
```

File: skyllh/core/analysis_utils.py [hunk 15]
```diff
@@ -157,8 +157,8 @@ def calculate_critical_ts_from_gamma(
     r = minimize(obj, x0, bounds=bounds)
     pars = r.x
     
-    interval = gamma.interval(1 - h0_ts_quantile, a=pars[0], scale=pars[1])
-    critical_ts = interval[1]
+    norm = alpha * (alpha_prime / (1-gamma.cdf(eta, a=pars[0], scale=pars[1]))
+    critical_ts = gamma.ppf(1 - 1./norm*h0_ts_quantile, a=pars[0], scale=pars[1])
```
Comment (HansN87) [inline skyllh/core/analysis_utils.py:161]: I think this only works for critical_ts > eta. 
We should think about checking/ensuring that his holds true.

File: skyllh/core/analysis_utils.py [hunk 16]
```diff
@@ -83,8 +87,82 @@ def calculate_pval_from_trials(
...
+
+    obj = lambda x: truncated_gamma_logpdf(x[0], x[1], eta=eta,
+                                           ts_above_eta=ts_eta,
+                                           N_above_eta=N_prime)
+    x0 = [0.75, 1.8]  # Initial values of function parameters.
+    bounds = [[0.1, 10], [0.1, 10]]  # Ranges for the minimization fitter.
+    r = minimize(obj, x0, bounds=bounds)
+    pars = r.x
+    
+    norm = alpha * (alpha_prime / (1-gamma.cdf(eta, a=pars[0], scale=pars[1])))
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:160]:
```suggestion
    norm = alpha*(alpha_prime/(1 - gamma.cdf(eta, a=pars[0], scale=pars[1])))
```

**Review by tomaskontrimas — CHANGES_REQUESTED** [review]
Spacing edit.

**Review by tomaskontrimas — APPROVED** [review]

File: skyllh/core/analysis_utils.py [hunk 17]
```diff
@@ -83,8 +87,88 @@ def calculate_pval_from_trials(
...
+
+    obj = lambda x: truncated_gamma_logpdf(x[0], x[1], eta=eta,
+                                           ts_above_eta=ts_eta,
+                                           N_above_eta=N_prime)
+    x0 = [0.75, 1.8]  # Initial values of function parameters.
+    bounds = [[0.1, 10], [0.1, 10]]  # Ranges for the minimization fitter.
+    r = minimize(obj, x0, bounds=bounds)
+    pars = r.x
+    
+    norm = alpha*(alpha_prime/(1 - gamma.cdf(eta, a=pars[0], scale=pars[1])))
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:160]: Survival function (also defined as 1 - cdf, but sf is sometimes more accurate).
```suggestion
    norm = alpha*(alpha_prime/gamma.sf(eta, a=pars[0], scale=pars[1]))
```

File: requirements.txt [hunk 18]
```diff
@@ -1,7 +1,10 @@
 astropy
 numpy
 scipy
+iminuit
 sphinx
 sphinx_rtd_theme
 nbsphinx
 pyyaml
+iminuit
```
Comment (tomaskontrimas) [inline requirements.txt:9]:
```suggestion
```

**Review by tomaskontrimas — APPROVED** [review]
Remove double requirement.

File: skyllh/core/analysis_utils.py [hunk 19]
```diff
@@ -6,6 +6,8 @@
 import numpy as np
 from numpy.lib import recfunctions as np_rfn
 import itertools
+import os.makedirs
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:9]:
```suggestion
from os import makedirs
```

File: skyllh/core/analysis_utils.py [hunk 20]
```diff
@@ -209,12 +316,22 @@ def estimate_mean_nsignal_for_ts_quantile(
...
+    if(h0_ts_quantile >= 1.e-3):
+        c = np.percentile(h0_ts_vals, (1 - h0_ts_quantile)*100)
+    else:
+        c = calculate_critical_ts_from_gamma(h0_ts_vals, h0_ts_quantile)
     logger.debug(
         'Critical ts value for bkg ts quantile %g: %e',
         h0_ts_quantile, c)
 
+    if(pathfilename is not None):
+        os.makedirs(os.path.dirname(pathfilename), exist_ok=True)
```
Comment (tomaskontrimas) [inline skyllh/core/analysis_utils.py:332]:
```suggestion
        makedirs(os.path.dirname(pathfilename), exist_ok=True)
```

**Review by tomaskontrimas — APPROVED** [review]
Fix broken import.

**Pull request merged**